Remove special case codegen for tril_indices/triu_indices.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33362 Standardize expanded TensorOptions representation in native_functions.yaml
* #33347 Update documentation on why _cudnn_init_dropout_state looks the way it is.
* **#33305 Remove special case codegen for tril_indices/triu_indices.**

The current TensorOptions code is written to exactly extract out
TensorOptions based on exact struct match, including default arguments.
That meant that tril_indices/triu_indices which had a different
default argument didn't match, and thus needed a special case.

I resolve this special case by instead replacing the explicit long
default argument with a None default argument, and then adjusting
the actual implementations to select the correct dtype when none
was specified.  I think the general rule I'm following here is that
it is always acceptable to replace an explicit default argument,
with a None argument (assuming the backend will compute it appropriately);
the documentation gets modestly worse, but everything that was
previously expressible continues to be expressible.  Maybe later
we should switch the default argument back to long, but for now
the simplification in code is worth it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19975411](https://our.internmc.facebook.com/intern/diff/D19975411)

ezyang committed:
Remove special case codegen for tril_indices/triu_indices.

The current TensorOptions code is written to exactly extract out
TensorOptions based on exact struct match, including default arguments.
That meant that tril_indices/triu_indices which had a different
default argument didn't match, and thus needed a special case.

I resolve this special case by instead replacing the explicit long
default argument with a None default argument, and then adjusting
the actual implementations to select the correct dtype when none
was specified.  I think the general rule I'm following here is that
it is always acceptable to replace an explicit default argument,
with a None argument (assuming the backend will compute it appropriately);
the documentation gets modestly worse, but everything that was
previously expressible continues to be expressible.  Maybe later
we should switch the default argument back to long, but for now
the simplification in code is worth it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

izdeby commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4359,12 +4359,12 @@
   use_c10_dispatcher: full
   variants: method, function
 
-- func: tril_indices(int row, int col, int offset=0, *, ScalarType? dtype=long, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Isn't it easier to have default values specified here rather somewhere in the code? native_functions.yaml is used as a reference doc often so it would make sense to have default types specified here.

ezyang replied:
In an ideal world, sure. But not if it's at the cost of an extremely painful series of hacks in the codegen to support this. Better to just do something marginally worse (but still 100% functional) and then maybe fix it later when codegen can handle it.

izdeby commented on aten/src/ATen/native/cuda/TensorFactories.cu:
```diff
@@ -321,7 +321,10 @@ void tril_indices_kernel(scalar_t * tensor,
-    int64_t row, int64_t col, int64_t offset, const TensorOptions& options) {
+    int64_t row, int64_t col, int64_t offset, const TensorOptions& options_) {
+
+  TensorOptions options = options_.dtype(options_.dtype_opt().value_or(scalarTypeToTypeMeta(kLong)));
```
nit: i would add a TODO what will say to change this once default values propagation will be implemented.

ezyang committed:
Update on "Remove special case codegen for tril_indices/triu_indices."


The current TensorOptions code is written to exactly extract out
TensorOptions based on exact struct match, including default arguments.
That meant that tril_indices/triu_indices which had a different
default argument didn't match, and thus needed a special case.

I resolve this special case by instead replacing the explicit long
default argument with a None default argument, and then adjusting
the actual implementations to select the correct dtype when none
was specified.  I think the general rule I'm following here is that
it is always acceptable to replace an explicit default argument,
with a None argument (assuming the backend will compute it appropriately);
the documentation gets modestly worse, but everything that was
previously expressible continues to be expressible.  Maybe later
we should switch the default argument back to long, but for now
the simplification in code is worth it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Remove special case codegen for tril_indices/triu_indices."


The current TensorOptions code is written to exactly extract out
TensorOptions based on exact struct match, including default arguments.
That meant that tril_indices/triu_indices which had a different
default argument didn't match, and thus needed a special case.

I resolve this special case by instead replacing the explicit long
default argument with a None default argument, and then adjusting
the actual implementations to select the correct dtype when none
was specified.  I think the general rule I'm following here is that
it is always acceptable to replace an explicit default argument,
with a None argument (assuming the backend will compute it appropriately);
the documentation gets modestly worse, but everything that was
previously expressible continues to be expressible.  Maybe later
we should switch the default argument back to long, but for now
the simplification in code is worth it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

VitalyFedyunin commented:
This one breaks pytorch_linux_backward_compatibility_check_test, can you please fix it
